tun2socks/lwip/device.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lwip
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+	lwipLib "github.com/eycorsican/go-tun2socks/core"
+)
+
+const packetMTU = 1500
+
+// LwIPTun2SocksDevice is an IPDevice that can translate IP packets to TCP/UDP traffic and vice versa. It uses the lwIP
+// library to perform the translation.
+//
+// LwIPTun2SocksDevice must be a singleton object due to limitations in lwIP.
+//
+// To use a LwIPTun2SocksDevice:
+//  1. Call NewTun2SocksDevice with two handlers for TCP and UDP traffic.
+//  2. Write IP packets to the device. The device will translate the IP packets to TCP/UDP traffic and send them to the
+//     appropriate handlers.
+//  3. Read from the device to get the TCP/UDP responses, which will be in IP packet format.
+type LwIPTun2SocksDevice struct {
+	tcp     *tcpHandler
+	udp     *udpHandler
+	stack   lwipLib.LWIPStack
+	outPipe *packetPipe
```

Comment:
Can't we use a `chan []byte` instead? Why not?
I think it's ok if reads and writes block on each other (It's the current behavior, I think?).

Reply:
I considered using `io.Pipe` before (which is just a simple `byte[]` channel) but decided not to use it because:

- It cannot handle "both `Read` and `WriteTo` at the same time" very well. for example, if you `Write` 5 packets, then `Read` 2, and call `WriteTo`; we expect that `WriteTo(w)` to flush the 3 remaining packets to `w`, and forward all subsequent `Write`s directly to `w`. If we use pipe, the 3rd `Write` is blocked forever because we only called 2 `Read`s, if we `close` it to unblock the pipe, then 3 remaining packets are lost.
- Also closing the pipe whenever `WriteTo` is called is semantically incorrect, because it will return `io.ErrClosedPipe` error to all `Read`s immediately, but actually the device is still open, so if someone calls `WriteTo`, we should block all other `Read`s and finally returns `EOF` to all `Read`s.
- lwIP will [call our `Write` periodically (as a callback function)](https://github.com/eycorsican/go-tun2socks/blob/master/core/output_export.go#L20). If our `Write` is blocking, not sure whether there will be any side effects.
- Also `io.Pipe`'s async implementation might access invalid memory. Take a look at the [lwIP's `output`](https://github.com/eycorsican/go-tun2socks/blob/master/core/output_export.go#L19-L20) function, the buffer passed to `Write(buf)` is a wrapper around a native memory, it might be released before the read go-routine actually kicks in and read the buffer.

Therefore I introduced this new class to handle these edge cases, it will always copy the native buffer to the internal buffer (if none `WriteTo` has been called). It also handles mixed `Read`/`Write`/`WriteTo` calls to ensure correct semantics of a `io.ReadWriter`. If caller has called `WriteTo`, no buffer will be allocated.

Reply:
Yes, this was my first design (synchronized version). You mentioned that we should actually block the write if no read occurs (so no ACK will be sent), I feel this should be the case (instead of the asynchronized version). Though we still need to add a comment mentioning that the buffer we passed to you will not be safe to use after the you return from the Write call.

Reply:
But one more thing which might complicate the scenario is that we need to handle the following calling sequence:

```
go t2s.Read()
go t2s.Read()
go t2s.WriteTo(w)
```

Though it is a valid calling sequence in a single thread app, but I still hope to cover that.